firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,311 @@
+	NVIC_Init_Struct.NVIC_IRQChannelPreemptionPriority = 5;
+	NVIC_Init_Struct.NVIC_IRQChannelSubPriority        = 0;
+	NVIC_Init(&NVIC_Init_Struct);
+
+	// NVIC Init for error interrupt
+	NVIC_Init_Struct.NVIC_IRQChannel                   = I2C_config.HWConfig->I2CERIRQ;
+	NVIC_Init_Struct.NVIC_IRQChannelCmd                = ENABLE;
+	NVIC_Init_Struct.NVIC_IRQChannelPreemptionPriority = 4;
+	NVIC_Init_Struct.NVIC_IRQChannelSubPriority        = 0;
+	NVIC_Init(&NVIC_Init_Struct);
+
+	// Set priority for ISR and enable ISR
+	NVIC_EnableIRQ(I2C_config.HWConfig->I2CEVIRQ);
+	NVIC_EnableIRQ(I2C_config.HWConfig->I2CERIRQ);
+}
+
+void I2C_setup(void)
+{
+	// I2C2 Init
+	RCCHelper_clockCmd(I2C_config.HWConfig->I2CPeriph, ENABLE);
+
+	/*
+	 * Reset I2C after a delay: After enabling the RCC clock, the BUSY bit in the SR2 register
+	 *                          became 1, which means the peripheral cannot start any data
+	 *                          transaction. According to STM32 documents and resources on
+	 *                          Google, this could be a silicon defect, and reseting the RCC
+	 *                          clock after a small delay is the current work-around.
+	 * (https://community.st.com/s/question/0D50X00009XkYxY/stm32f105rc-i2c2-always-busy)
+	 *
+	 * 15 is the minimum value of delay that ensures I2C's functioning for now.
+	 * If the BUSY bit cannot be reset, increase the delay time.
+	 * DO NOT check BUSY bit in SR2 in the code!!! It will then be set FOREVER!!!
+	 */
+	int i = 0;
+	while (i < 15){
+		i++;
+	}
+	RCCHelper_clockCmd(I2C_config.HWConfig->I2CPeriph, ENABLE);
+	RCCHelper_clockCmd(I2C_config.HWConfig->I2CPeriph, DISABLE);
+
+	I2C_InitTypeDef I2C_Init_Struct;
+	I2C_Init_Struct.I2C_ClockSpeed          = 50000; // Any requirement for I2C clock speed?
+	I2C_Init_Struct.I2C_Ack                 = I2C_Ack_Enable;
+	I2C_Init_Struct.I2C_AcknowledgedAddress = I2C_AcknowledgedAddress_7bit;
+	I2C_Init_Struct.I2C_DutyCycle           = I2C_DutyCycle_2;
+	I2C_Init_Struct.I2C_Mode                = I2C_Mode_I2C;
+	I2C_Init_Struct.I2C_OwnAddress1         = 0x0;
+	I2C_Init(I2C_config.HWConfig->I2CPeriph, &I2C_Init_Struct);
+
+	// Enable event interrupts and buffer interrupts
+	I2C_ITConfig(I2C_config.HWConfig->I2CPeriph, I2C_IT_EVT, ENABLE);
+	I2C_ITConfig(I2C_config.HWConfig->I2CPeriph, I2C_IT_BUF, ENABLE);
+	I2C_ITConfig(I2C_config.HWConfig->I2CPeriph, I2C_IT_ERR, ENABLE);
+
+	// Enable I2C2
+	// This doesn't work, PE cannot be set
+	//I2C_Cmd(I2C_config.HWConfig->I2CPeriph, ENABLE);
+	I2C_config.HWConfig->I2CPeriph->CR1 |= I2C_CR1_PE;
+}
+
+bool I2C_send(I2C_channel_E channel, const uint8_t * data, const uint8_t length)
+{
+	bool task_delivery = false;
+
+	// Check the validity of message
+	if((data != NULL) && (length != 0) && (length <= _I2C_BUFFER_SIZE))
+	{
+		// Check the availability of I2C
+		if ((data_length == 0) && (I2C_state == I2C_STATE_IDLE)) {
+
+			memcpy(I2C_send_data.Buffer, data, length);

```

Comment:
There is a circBuffer1D module you can use instead of a buffer like this if you want.